python/dgl/sampling/negative.py

```diff
@@ -0,0 +1,72 @@
+"""Negative sampling APIs"""
+
+from .._ffi.function import _init_api
+from .. import backend as F
+
+__all__ = [
+    'global_uniform_negative_sampling']
+
+def global_uniform_negative_sampling(
+        g, num_samples, exclude_self_loops=True, unique=True, etype=None, num_trials=3,
+        redundancy=1.3):
+    """Performs negative sampling, which generate source-destination pairs such that
+    edges with the given type do not exist.
+
+    Specifically, this function takes in an edge type and a number of samples.  It
+    returns two tensors ``src`` and ``dst``, the former in the range of ``[0, num_src)``
+    and the latter in the range of ``[0, num_dst)``, where ``num_src`` and ``num_dst``
+    represents the number of nodes with the source and destination node type respectively.
+    It guarantees that no edge will exist between the corresponding pairs of ``src``
+    with the source node type and ``dst`` with the destination node type.
+
+    .. note::
+
+       This function uses rejection sampling, and may not always return the same number
+       of negative examples as the given :attr:`num_samples`.  This is more likely to
+       happen when a graph is so small or dense that not many unique negative examples exist.
+
+    Parameters
+    ----------
+    g : DGLGraph
+        The graph.
+    num_samples : int
+        The number of negative examples to generate.
+    exclude_self_loops : bool, optional
+        Whether to exclude self-loops from the negative examples.  Only impacts the
+        edge types whose source and destination node types are the same.
+
+        Default: True.
+    unique : bool, optional
+        Whether to sample unique negative examples.  Setting it to False will make things
+        faster.  (Default: True)
+    etype : str or tuple of str, optional
+        The edge type.  Can be omitted if the graph only has one edge type.
```

Comment:
I don't think this is a common use case as the task will be a lot easier than negative examples with the same types; most of the time you will only need to know which types every node belongs to.

@zjost If you agree with Minjie then I'll consider adding a homogenized negative sampler later.

Reply:
Is the `etype` needed because we use it to construct a heterogeneous graph for the negative samples?  If that's the case, is the question whether or not we should sample the edge type or keep it static?

A follow-up question:  given `etype`, let's say you generate a negative sample, `(u, v)`.  If `(u, v)` shares an edge in the real graph, but it's of a different type than `etype`, do you still exclude it, or do you only remove positive pairs with `etype` edges?

Reply:
> Is the `etype` needed because we use it to construct a heterogeneous graph for the negative samples?

Yes.

> If that's the case, is the question whether or not we should sample the edge type or keep it static?

Yes, or "should we sample the edge type for negative examples as well".  I think those negative examples will be too easy.

> If `(u, v)` shares an edge in the real graph, but it's of a different type than `etype`, do you still exclude it, or do you only remove positive pairs with `etype` edges?

I only remove positive pairs with `etype` edges.